src/libraries/System.Private.CoreLib/src/System/Resources/ResourceReader.Core.cs

```diff
@@ -69,19 +65,34 @@ private object DeserializeObject(int typeIndex)
 
             if (Volatile.Read(ref _binaryFormatter) is null)
             {
-                if (!InitializeBinaryFormatterLocal())
+                try
                 {
-                    // Trimming took away the BinaryFormatter implementation and we can't call into it.
-                    // We'll throw an exception with the same text that BinaryFormatter would have thrown
-                    // had we been able to call into it. Keep this resource string in sync with the same
-                    // resource from the Formatters assembly.
-                    throw new NotSupportedException(SR.BinaryFormatter_SerializationDisallowed);
+                    if (!InitializeBinaryFormatterLocal())
+                    {
+                        // Trimming took away the BinaryFormatter implementation and we can't call into it.
+                        // We'll throw an exception with the same text that BinaryFormatter would have thrown
+                        // had we been able to call into it. Keep this resource string in sync with the same
+                        // resource from the Formatters assembly.
+                        throw new NotSupportedException(SR.BinaryFormatter_SerializationDisallowed);
+                    }
+                }
+                catch (Exception ex) when (ex is TypeLoadException or MissingMethodException)
+                {
+                    throw new NotSupportedException(SR.BinaryFormatter_SerializationDisallowed, ex);
```

Comment:
```suggestion
                    InitializeBinaryFormatterLocal()
                }
                catch (Exception ex) when (ex is TypeLoadException or MissingMethodException)
                {
                    // Trimming took away the BinaryFormatter implementation and we can't call into it.
                    // We'll throw an exception with the same text that BinaryFormatter would have thrown
                    // had we been able to call into it. Keep this resource string in sync with the same
                    // resource from the Formatters assembly.
                    throw new NotSupportedException(SR.BinaryFormatter_SerializationDisallowed, ex);
```
I am not sure why we need this try/catch in the first place. Trimming won't "take away" the BinaryFormatter implementation. This may be a left-over from times when the trimmer was not smart enough.

Reply:
We have a special ILLink substitution to ensure that BinaryFormatter gets linked away. I have over-complicated this with the exception handling (from before I understood how the substitution works). I'll clean this up.